File: legacy/packages/parser/src/declarations.ts
import type {
  AgentDecl,
  ConfigItem,
  Constructor,
  DataDecl,
  Declaration,
  FieldDef,
  FieldModifier,
  FunctionDecl,
  FunctionEquation,
  InstanceDecl,
  MacroDecl,
  ModelDecl,
  Program,
  SourceSpec,
} from "@moon/ast";
import { ParseError, type TokenStream } from "./token-stream";
import { parseTypeSpec } from "./types";
import { parseExpression, parseDoBlock, parseDoStatementFixed, parseConfigItems } from "./expressions";
import { parsePattern } from "./patterns";

export function parseProgram(ts: TokenStream): Program {
  const start = ts.peek();
  const declarations: Declaration[] = [];

  while (!ts.at("EOF")) {
    if (ts.skip("SEMI")) continue;
    declarations.push(parseDeclaration(ts));
  }

  return { declarations, span: ts.makeSpan(start, ts.last()) };
}

function parseDeclaration(ts: TokenStream): Declaration {
  const start = ts.peek();

  if (ts.at("KW_IMPORT")) return parseImport(ts);
  if (ts.at("KW_MODEL")) return { kind: "Model", decl: parseModelDecl(ts), span: ts.makeSpan(start, ts.last()) };
  if (ts.at("KW_AGENT")) return { kind: "Agent", decl: parseAgentDecl(ts), span: ts.makeSpan(start, ts.last()) };
  if (ts.at("KW_DATA")) return { kind: "Data", decl: parseDataDecl(ts), span: ts.makeSpan(start, ts.last()) };
  if (ts.at("KW_INSTANCE")) return { kind: "Instance", decl: parseInstanceDecl(ts), span: ts.makeSpan(start, ts.last()) };
  if (ts.at("KW_MACRO")) return { kind: "Macro", decl: parseMacroDecl(ts), span: ts.makeSpan(start, ts.last()) };
  if (ts.at("IDENT") || ts.at("KW_PURE")) {
    return { kind: "Function", decl: parseFunctionDecl(ts), span: ts.makeSpan(start, ts.last()) };
  }

  throw new ParseError("Expected declaration", start.line, start.column);
}

function parseDeclName(ts: TokenStream): string {
  if (ts.at("KW_PURE")) return ts.advance().value ?? "pure";
  return ts.expect("IDENT").value!;
}

function isDeclNameStart(ts: TokenStream): boolean {
  return ts.at("IDENT") || ts.at("KW_PURE");
}

function parseImport(ts: TokenStream): Declaration {
  const start = ts.expect("KW_IMPORT");
  const path: string[] = [ts.expect("IDENT").value!];
  while (ts.skip("FIELD_DOT")) {
    path.push(ts.expect("IDENT").value!);
  }
  let alias: string | undefined;
  if (ts.skip("KW_AS")) {
    alias = ts.expect("IDENT").value;
  }
  return { kind: "Import", path, alias, span: ts.makeSpan(start, ts.last()) };
}

export function parseModelDecl(ts: TokenStream): ModelDecl {
  const start = ts.expect("KW_MODEL");
  const name = ts.expect("IDENT").value!;
  const typeParams = parseTypeParams(ts);

  let implements_: string | undefined;
  if (ts.skip("KW_IMPLEMENTS")) {
    implements_ = ts.expect("IDENT").value;
  }

  const fields: FieldDef[] = [];

  if (ts.skip("KW_WHERE")) {
    if (ts.at("LBRACE")) ts.advance();
    while (!ts.at("RBRACE") && !ts.at("EOF")) {
      if (ts.skip("SEMI")) continue;
      fields.push(parseFieldDef(ts));
    }
    if (ts.at("RBRACE")) ts.advance();
  }

  return { name, typeParams, implements: implements_, fields, span: ts.makeSpan(start, ts.last()) };
}

function parseFieldDef(ts: TokenStream): FieldDef {
  const start = ts.peek();
  const name = ts.expect("IDENT").value!;
  ts.expect("COLON");
  ts.expect("COLON");
  const type = parseTypeSpec(ts);
  const modifiers = parseFieldModifiers(ts);
  return { name, type, modifiers, span: ts.makeSpan(start, ts.last()) };
}

function parseFieldModifiers(ts: TokenStream): FieldModifier[] {
  const mods: FieldModifier[] = [];

  while (true) {
    if (ts.at("KW_CONSTRAINT")) {
      const s = ts.advance();
      ts.expect("COLON");
      const expr = parseExpression(ts);
      mods.push({ kind: "Constraint", expr, span: ts.makeSpan(s, ts.last()) });
      continue;
    }
    if (ts.at("KW_DEFAULT")) {
      const s = ts.advance();
      ts.expect("COLON");
      const expr = parseExpression(ts);
      mods.push({ kind: "Default", expr, span: ts.makeSpan(s, ts.last()) });
      continue;
    }
    if (ts.at("KW_FETCHED")) {
      const s = ts.advance();
      ts.expect("KW_FROM");
      const sources = parseSourceSpecs(ts);
      mods.push({ kind: "FetchedFrom", sources, span: ts.makeSpan(s, ts.last()) });
      continue;
    }
    if (ts.at("KW_OPTIONAL")) {
      const s = ts.advance();
      mods.push({ kind: "Optional", span: ts.makeSpan(s, s) });
      continue;
    }
    break;
  }

  return mods;
}

function parseSourceSpecs(ts: TokenStream): SourceSpec[] {
  const specs: SourceSpec[] = [];
  do {
    const start = ts.peek();
    const source = ts.expect("IDENT").value!;
    ts.expect("COLON");
    const field = ts.expect("IDENT").value!;
    specs.push({ source, field, span: ts.makeSpan(start, ts.last()) });
  } while (ts.skip("PIPE"));
  return specs;
}

export function parseAgentDecl(ts: TokenStream): AgentDecl {
  const start = ts.expect("KW_AGENT");
  const name = ts.expect("IDENT").value!;
  const typeParams = parseTypeParams(ts);

  ts.expect("COLON");
  ts.expect("COLON");
  const type = parseTypeSpec(ts);

  let routesTo: string | undefined;
  if (ts.skip("KW_ROUTES_TO")) {
    routesTo = ts.expect("IDENT").value;
  }

  const config: ConfigItem[] = [];

  if (ts.skip("KW_WHERE")) {
    if (ts.at("LBRACE")) ts.advance();
    config.push(...parseConfigBlock(ts));
    if (ts.at("RBRACE")) ts.advance();
  } else {
    if (ts.at("LBRACE")) ts.advance();
    if (ts.at("IDENT") && ts.peek(1).kind === "COLON") {
      config.push(...parseConfigBlock(ts));
    }
    if (ts.at("RBRACE")) ts.advance();
  }

  return { name, typeParams, type, routesTo, config, span: ts.makeSpan(start, ts.last()) };
}

function parseConfigBlock(ts: TokenStream): ConfigItem[] {
  const items: ConfigItem[] = [];
  while (!ts.at("RBRACE") && !ts.at("EOF")) {
    if (ts.skip("SEMI")) continue;
    if (!(ts.at("IDENT") && ts.peek(1).kind === "COLON")) break;
    const keyStart = ts.peek();
    const key = ts.advance().value!;
    ts.expect("COLON");
    const value = parseExpression(ts);
    items.push({ key, value, span: ts.makeSpan(keyStart, ts.last()) });
    if (ts.skip("SEMI")) continue;
  }
  return items;
}

function parseDataDecl(ts: TokenStream): DataDecl {
  const start = ts.expect("KW_DATA");
  const name = ts.expect("IDENT").value!;
  const typeParams = parseTypeParams(ts);
  ts.expect("EQUALS");

  const constructors: Constructor[] = [parseConstructor(ts)];
  while (ts.skip("PIPE")) {
    constructors.push(parseConstructor(ts));
  }

  return { name, typeParams, constructors, span: ts.makeSpan(start, ts.last()) };
}

function parseConstructor(ts: TokenStream): Constructor {
  const start = ts.peek();
  const name = ts.expect("IDENT").value!;

  if (ts.at("LPAREN")) {
    ts.advance();
    const types: ReturnType<typeof parseTypeSpec>[] = [];
    if (!ts.at("RPAREN")) {
      types.push(parseTypeSpec(ts));
      while (ts.skip("COMMA")) {
        types.push(parseTypeSpec(ts));
      }
    }
    const end = ts.expect("RPAREN");
    return {
      name,
      args: { kind: "Positional", types, span: ts.makeSpan(start, end) },
      span: ts.makeSpan(start, end),
    };
  }

  if (ts.at("LBRACE")) {
    ts.advance();
    const fields: { name: string; type: ReturnType<typeof parseTypeSpec>; span: typeof start extends infer _ ? { start: { line: number; column: number; offset: number }; end: { line: number; column: number; offset: number } } : never }[] = [];
    while (!ts.at("RBRACE")) {
      if (ts.skip("SEMI") || ts.skip("COMMA")) continue;
      const fs = ts.peek();
      const fname = ts.expect("IDENT").value!;
      ts.expect("COLON");
      ts.expect("COLON");
      const ftype = parseTypeSpec(ts);
      fields.push({ name: fname, type: ftype, span: ts.makeSpan(fs, ts.last()) });
      if (ts.skip("COMMA")) continue;
      if (ts.skip("SEMI")) continue;
    }
    const end = ts.expect("RBRACE");
    return {
      name,
      args: { kind: "Record", fields, span: ts.makeSpan(start, end) },
      span: ts.makeSpan(start, end),
    };
  }

  return { name, span: ts.makeSpan(start, ts.last()) };
}

function parseInstanceDecl(ts: TokenStream): InstanceDecl {
  const start = ts.expect("KW_INSTANCE");
  const className = ts.expect("IDENT").value!;
  ts.expect("KW_FOR");
  const type = parseTypeSpec(ts);
  const typeParams = parseTypeParams(ts);

  const functions: FunctionDecl[] = [];
  ts.expect("KW_WHERE");
  if (ts.at("LBRACE")) ts.advance();

  while (!ts.at("RBRACE") && !ts.at("EOF")) {
    if (ts.skip("SEMI")) continue;
    functions.push(parseFunctionDecl(ts));
  }
  if (ts.at("RBRACE")) ts.advance();

  return { className, type, typeParams, functions, span: ts.makeSpan(start, ts.last()) };
}

function parseMacroDecl(ts: TokenStream): MacroDecl {
  const start = ts.expect("KW_MACRO");
  const name = ts.expect("IDENT").value!;
  const typeParams = parseTypeParams(ts);
  const params = parseParamList(ts);
  ts.expect("EQUALS");
  const body = parseDoBlock(ts);
  return { name, typeParams, params, body, span: ts.makeSpan(start, ts.last()) };
}

function parseFunctionDecl(ts: TokenStream): FunctionDecl {
  const start = ts.peek();
  let signature: FunctionDecl["signature"];

  if (isDeclNameStart(ts) && ts.peek(1).kind === "COLON" && ts.peek(2).kind === "COLON") {
    const name = parseDeclName(ts);
    ts.expect("COLON");
    ts.expect("COLON");
    const type = parseTypeSpec(ts);
    signature = { name, type, span: ts.makeSpan(start, ts.last()) };
  }

  const equations: FunctionEquation[] = [parseFunctionEquation(ts, signature?.name)];

  while (isDeclNameStart(ts) && !isTypeSignatureStart(ts)) {
    const nextName = ts.peek().value ?? "pure";
    const expected = signature?.name ?? equations[0]!.name;
    if (nextName !== expected) break;
    equations.push(parseFunctionEquation(ts, expected));
  }

  return { signature, equations, span: ts.makeSpan(start, ts.last()) };
}

function isTypeSignatureStart(ts: TokenStream): boolean {
  return isDeclNameStart(ts) && ts.peek(1).kind === "COLON" && ts.peek(2).kind === "COLON";
}

function parseFunctionEquation(ts: TokenStream, expectedName?: string): FunctionEquation {
  const start = ts.peek();
  const name = parseDeclName(ts);
  if (expectedName && name !== expectedName) {
    throw new ParseError(`Expected equation for ${expectedName}, got ${name}`, start.line, start.column);
  }

  const patterns: ReturnType<typeof parsePattern>[] = [];
  while (isPatternContinuer(ts)) {
    patterns.push(parsePattern(ts, { allowApp: false }));
  }

  ts.expect("EQUALS");

  let body: ReturnType<typeof parseExpression> | ReturnType<typeof parseDoBlock>;
  if (ts.at("KW_DO")) {
    body = parseDoBlock(ts);
  } else if (ts.at("KW_AGENT")) {
    body = { kind: "Agent", decl: parseAgentDecl(ts), span: ts.makeSpan(start, ts.last()) } as ReturnType<typeof parseExpression>;
  } else {
    body = parseExpression(ts);
  }

  return { name, patterns, body, span: ts.makeSpan(start, ts.last()) };
}

function isPatternContinuer(ts: TokenStream): boolean {
  if (ts.at("EQUALS") || ts.at("COLON")) return false;
  return ts.check("IDENT", "LPAREN", "LBRACKET", "STRING", "INT", "FLOAT", "KW_TRUE", "KW_FALSE");
}

function parseTypeParams(ts: TokenStream): string[] {
  const params: string[] = [];
  while (ts.at("IDENT") && !isDeclBoundary(ts)) {
    const next = ts.peek(1).kind;
    if (next === "COLON" && ts.peek(2).kind === "COLON") {
      params.push(ts.advance().value!);
      break;
    }
    if (next === "EQUALS") {
      params.push(ts.advance().value!);
      break;
    }
    params.push(ts.advance().value!);
  }
  return params;
}

function isDeclBoundary(ts: TokenStream): boolean {
  return ts.check("KW_WHERE", "KW_IMPLEMENTS", "KW_ROUTES_TO", "EQUALS", "EOF", "SEMI", "RBRACE");
}

function parseParamList(ts: TokenStream): { name: string; type: ReturnType<typeof parseTypeSpec> }[] {
  if (!ts.at("LPAREN")) return [];
  ts.advance();
  const params: { name: string; type: ReturnType<typeof parseTypeSpec> }[] = [];
  if (!ts.at("RPAREN")) {
    const pname = ts.expect("IDENT").value!;
    ts.expect("COLON");
    ts.expect("COLON");
    params.push({ name: pname, type: parseTypeSpec(ts) });
    while (ts.skip("COMMA")) {
      const n = ts.expect("IDENT").value!;
      ts.expect("COLON");
      ts.expect("COLON");
      params.push({ name: n, type: parseTypeSpec(ts) });
    }
  }
  ts.expect("RPAREN");
  return params;
}

File: Sources/MoonParser/Declarations.swift
import MoonAST
import MoonLexer

public func parseProgram(_ ts: TokenStream) throws -> Program {
    let start = ts.peek()
    var declarations: [Declaration] = []

    while !ts.at(.eof) {
        if ts.skip(.semi) { continue }
        declarations.append(try parseDeclaration(ts))
    }

    return Program(declarations: declarations, span: ts.makeSpan(start: start, end: ts.last()))
}

private func parseDeclaration(_ ts: TokenStream) throws -> Declaration {
    let start = ts.peek()

    if ts.at(.kwImport) {
        return try parseImport(ts)
    }
    if ts.at(.kwModel) {
        let decl = try parseModelDecl(ts)
        return .model(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }
    if ts.at(.kwAgent) {
        let decl = try parseAgentDecl(ts)
        return .agent(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }
    if ts.at(.kwData) {
        let decl = try parseDataDecl(ts)
        return .data(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }
    if ts.at(.kwInstance) {
        let decl = try parseInstanceDecl(ts)
        return .instance(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }
    if ts.at(.kwMacro) {
        let decl = try parseMacroDecl(ts)
        return .macro(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }
    if ts.at(.ident) || ts.at(.kwPure) {
        let decl = try parseFunctionDecl(ts)
        return .function(decl: decl, span: ts.makeSpan(start: start, end: ts.last()))
    }

    throw ParseError("Expected declaration", line: start.line, column: start.column)
}

private func parseImport(_ ts: TokenStream) throws -> Declaration {
    let start = try ts.expect(.kwImport)
    var path = [try ts.expect(.ident).value ?? ""]
    while ts.skip(.fieldDot) {
        path.append(try ts.expect(.ident).value ?? "")
    }
    var alias: String?
    if ts.skip(.kwAs) {
        alias = try ts.expect(.ident).value
    }
    return .importDecl(path: path, alias: alias, span: ts.makeSpan(start: start, end: ts.last()))
}

public func parseModelDecl(_ ts: TokenStream) throws -> ModelDecl {
    let start = try ts.expect(.kwModel)
    let name = try ts.expect(.ident).value ?? ""
    let typeParams = try parseTypeParams(ts)

    var implements: String?
    if ts.skip(.kwImplements) {
        implements = try ts.expect(.ident).value
    }

    var fields: [FieldDef] = []

    if ts.skip(.kwWhere) {
        if ts.at(.lbrace) { _ = ts.advance() }
        while !ts.at(.rbrace), !ts.at(.eof) {
            if ts.skip(.semi) { continue }
            fields.append(try parseFieldDef(ts))
        }
        if ts.at(.rbrace) { _ = ts.advance() }
    }

    return ModelDecl(
        name: name,
        typeParams: typeParams,
        implements: implements,
        fields: fields,
        span: ts.makeSpan(start: start, end: ts.last())
    )
}

private func parseFieldDef(_ ts: TokenStream) throws -> FieldDef {
    let start = ts.peek()
    let name = try ts.expect(.ident).value ?? ""
    try ts.expect(.colon)
    try ts.expect(.colon)
    let type = try parseTypeSpec(ts)
    let modifiers = try parseFieldModifiers(ts)
    return FieldDef(name: name, type: type, modifiers: modifiers, span: ts.makeSpan(start: start, end: ts.last()))
}

private func parseFieldModifiers(_ ts: TokenStream) throws -> [FieldModifier] {
    var mods: [FieldModifier] = []

    while true {
        if ts.at(.kwConstraint) {
            let s = ts.advance()
            try ts.expect(.colon)
            let expr = try parseExpression(ts)
            mods.append(.constraint(expr: expr, span: ts.makeSpan(start: s, end: ts.last())))
            continue
        }
        if ts.at(.kwDefault) {
            let s = ts.advance()
            try ts.expect(.colon)
            let expr = try parseExpression(ts)
            mods.append(.defaultValue(expr: expr, span: ts.makeSpan(start: s, end: ts.last())))
            continue
        }
        if ts.at(.kwFetched) {
            let s = ts.advance()
            try ts.expect(.kwFrom)
            let sources = try parseSourceSpecs(ts)
            mods.append(.fetchedFrom(sources: sources, span: ts.makeSpan(start: s, end: ts.last())))
            continue
        }
        if ts.at(.kwOptional) {
            let s = ts.advance()
            mods.append(.optional(span: ts.makeSpan(start: s, end: s)))
            continue
        }
        break
    }

    return mods
}

private func parseSourceSpecs(_ ts: TokenStream) throws -> [SourceSpec] {
    var specs: [SourceSpec] = []
    repeat {
        let start = ts.peek()
        let source = try ts.expect(.ident).value ?? ""
        try ts.expect(.colon)
        let field = try ts.expect(.ident).value ?? ""
        specs.append(SourceSpec(source: source, field: field, span: ts.makeSpan(start: start, end: ts.last())))
    } while ts.skip(.pipe)
    return specs
}

public func parseAgentDecl(_ ts: TokenStream) throws -> AgentDecl {
    let start = try ts.expect(.kwAgent)
    let name = try ts.expect(.ident).value ?? ""
    let typeParams = try parseTypeParams(ts)

    try ts.expect(.colon)
    try ts.expect(.colon)
    let type = try parseTypeSpec(ts)

    var routesTo: String?
    if ts.skip(.kwRoutesTo) {
        routesTo = try ts.expect(.ident).value
    }

    var config: [ConfigItem] = []

    if ts.skip(.kwWhere) {
        if ts.at(.lbrace) { _ = ts.advance() }
        config.append(contentsOf: try parseConfigBlock(ts))
        if ts.at(.rbrace) { _ = ts.advance() }
    } else {
        if ts.at(.lbrace) { _ = ts.advance() }
        if ts.at(.ident), ts.peek(offset: 1).kind == .colon {
            config.append(contentsOf: try parseConfigBlock(ts))
        }
        if ts.at(.rbrace) { _ = ts.advance() }
    }

    return AgentDecl(
        name: name,
        typeParams: typeParams,
        type: type,
        routesTo: routesTo,
        config: config,
        span: ts.makeSpan(start: start, end: ts.last())
    )
}

private func parseConfigBlock(_ ts: TokenStream) throws -> [ConfigItem] {
    var items: [ConfigItem] = []
    while !ts.at(.rbrace), !ts.at(.eof) {
        if ts.skip(.semi) { continue }
        guard ts.at(.ident), ts.peek(offset: 1).kind == .colon else { break }
        let keyStart = ts.peek()
        let key = ts.advance().value ?? ""
        try ts.expect(.colon)
        let value = try parseExpression(ts)
        items.append(ConfigItem(key: key, value: value, span: ts.makeSpan(start: keyStart, end: ts.last())))
        if ts.skip(.semi) { continue }
    }
    return items
}

private func parseDataDecl(_ ts: TokenStream) throws -> DataDecl {
    let start = try ts.expect(.kwData)
    let name = try ts.expect(.ident).value ?? ""
    let typeParams = try parseTypeParams(ts)
    try ts.expect(.equals)

    var constructors = [try parseConstructor(ts)]
    while ts.skip(.pipe) {
        constructors.append(try parseConstructor(ts))
    }

    return DataDecl(
        name: name,
        typeParams: typeParams,
        constructors: constructors,
        span: ts.makeSpan(start: start, end: ts.last())
    )
}

private func parseConstructor(_ ts: TokenStream) throws -> Constructor {
    let start = ts.peek()
    let name = try ts.expect(.ident).value ?? ""

    if ts.at(.lparen) {
        _ = ts.advance()
        var types: [TypeSpec] = []
        if !ts.at(.rparen) {
            types.append(try parseTypeSpec(ts))
            while ts.skip(.comma) {
                types.append(try parseTypeSpec(ts))
            }
        }
        let end = try ts.expect(.rparen)
        return Constructor(
            name: name,
            args: ConstructorArgs.positional(types: types, span: ts.makeSpan(start: start, end: end)),
            span: ts.makeSpan(start: start, end: end)
        )
    }

    if ts.at(.lbrace) {
        _ = ts.advance()
        var fields: [RecordFieldType] = []
        while !ts.at(.rbrace) {
            if ts.skip(.semi) || ts.skip(.comma) { continue }
            let fs = ts.peek()
            let fname = try ts.expect(.ident).value ?? ""
            try ts.expect(.colon)
            try ts.expect(.colon)
            let ftype = try parseTypeSpec(ts)
            fields.append(RecordFieldType(name: fname, type: ftype, span: ts.makeSpan(start: fs, end: ts.last())))
            if ts.skip(.comma) { continue }
            if ts.skip(.semi) { continue }
        }
        let end = try ts.expect(.rbrace)
        return Constructor(
            name: name,
            args: ConstructorArgs.record(fields: fields, span: ts.makeSpan(start: start, end: end)),
            span: ts.makeSpan(start: start, end: end)
        )
    }

    return Constructor(name: name, args: nil, span: ts.makeSpan(start: start, end: ts.last()))
}

private func parseInstanceDecl(_ ts: TokenStream) throws -> InstanceDecl {
    let start = try ts.expect(.kwInstance)
    let className = try ts.expect(.ident).value ?? ""
    try ts.expect(.kwFor)
    let type = try parseTypeSpec(ts)
    let typeParams = try parseTypeParams(ts)

    var functions: [FunctionDecl] = []
    try ts.expect(.kwWhere)
    if ts.at(.lbrace) { _ = ts.advance() }

    while !ts.at(.rbrace), !ts.at(.eof) {
        if ts.skip(.semi) { continue }
        functions.append(try parseFunctionDecl(ts))
    }
    if ts.at(.rbrace) { _ = ts.advance() }

    return InstanceDecl(
        className: className,
        type: type,
        typeParams: typeParams,
        functions: functions,
        span: ts.makeSpan(start: start, end: ts.last())
    )
}

private func parseMacroDecl(_ ts: TokenStream) throws -> MacroDecl {
    let start = try ts.expect(.kwMacro)
    let name = try ts.expect(.ident).value ?? ""
    let typeParams = try parseTypeParams(ts)
    let params = try parseParamList(ts)
    try ts.expect(.equals)
    let body = try parseDoBlock(ts)
    return MacroDecl(
        name: name,
        typeParams: typeParams,
        params: params,
        body: body,
        span: ts.makeSpan(start: start, end: ts.last())
    )
}

private func parseFunctionDecl(_ ts: TokenStream) throws -> FunctionDecl {
    let start = ts.peek()
    var signature: FunctionSignature?

    if isDeclNameStart(ts), ts.peek(offset: 1).kind == .colon, ts.peek(offset: 2).kind == .colon {
        let name = try parseDeclName(ts)
        try ts.expect(.colon)
        try ts.expect(.colon)
        let type = try parseTypeSpec(ts)
        signature = FunctionSignature(name: name, type: type, span: ts.makeSpan(start: start, end: ts.last()))
    }

    var equations = [try parseFunctionEquation(ts, expectedName: signature?.name)]
    while isDeclNameStart(ts), !isTypeSignatureStart(ts) {
        let nextName = ts.peek().value ?? "pure"
        let expected = signature?.name ?? equations[0].name
        if nextName != expected { break }
        equations.append(try parseFunctionEquation(ts, expectedName: expected))
    }

    return FunctionDecl(signature: signature, equations: equations, span: ts.makeSpan(start: start, end: ts.last()))
}

private func isDeclNameStart(_ ts: TokenStream) -> Bool {
    ts.at(.ident) || ts.at(.kwPure)
}

private func parseDeclName(_ ts: TokenStream) throws -> String {
    if ts.at(.kwPure) {
        return ts.advance().value ?? "pure"
    }
    return try ts.expect(.ident).value ?? ""
}

private func isTypeSignatureStart(_ ts: TokenStream) -> Bool {
    isDeclNameStart(ts) && ts.peek(offset: 1).kind == .colon && ts.peek(offset: 2).kind == .colon
}

private func parseFunctionEquation(_ ts: TokenStream, expectedName: String?) throws -> FunctionEquation {
    let start = ts.peek()
    let name = try parseDeclName(ts)
    if let expectedName, name != expectedName {
        throw ParseError("Expected equation for \(expectedName), got \(name)", line: start.line, column: start.column)
    }

    var patterns: [Pattern] = []
    while isPatternContinuer(ts) {
        patterns.append(try parsePattern(ts, options: ParsePatternOptions(allowApp: false)))
    }

    try ts.expect(.equals)

    let body: EquationBody
    if ts.at(.kwDo) {
        body = .doBlock(try parseDoBlock(ts))
    } else if ts.at(.kwAgent) {
        let decl = try parseAgentDecl(ts)
        body = .expression(.agent(decl: decl, span: ts.makeSpan(start: start, end: ts.last())))
    } else {
        body = .expression(try parseExpression(ts))
    }

    return FunctionEquation(name: name, patterns: patterns, body: body, span: ts.makeSpan(start: start, end: ts.last()))
}

private func isPatternContinuer(_ ts: TokenStream) -> Bool {
    if ts.at(.equals) || ts.at(.colon) { return false }
    return ts.check(.ident, .lparen, .lbracket, .string, .int, .float, .kwTrue, .kwFalse)
}

private func parseTypeParams(_ ts: TokenStream) throws -> [String] {
    var params: [String] = []
    while ts.at(.ident), !isDeclBoundary(ts) {
        let next = ts.peek(offset: 1).kind
        if next == .colon, ts.peek(offset: 2).kind == .colon {
            params.append(ts.advance().value ?? "")
            break
        }
        if next == .equals {
            params.append(ts.advance().value ?? "")
            break
        }
        params.append(ts.advance().value ?? "")
    }
    return params
}

private func isDeclBoundary(_ ts: TokenStream) -> Bool {
    ts.check(.kwWhere, .kwImplements, .kwRoutesTo, .equals, .eof, .semi, .rbrace)
}

private func parseParamList(_ ts: TokenStream) throws -> [ParamDef] {
    guard ts.at(.lparen) else { return [] }
    _ = ts.advance()
    var params: [ParamDef] = []
    if !ts.at(.rparen) {
        let pname = try ts.expect(.ident).value ?? ""
        try ts.expect(.colon)
        try ts.expect(.colon)
        params.append(ParamDef(name: pname, type: try parseTypeSpec(ts)))
        while ts.skip(.comma) {
            let n = try ts.expect(.ident).value ?? ""
            try ts.expect(.colon)
            try ts.expect(.colon)
            params.append(ParamDef(name: n, type: try parseTypeSpec(ts)))
        }
    }
    try ts.expect(.rparen)
    return params
}